File: 2026/08-August/01-Saturday/solution.cs
public class Solution
{
    private int?[,] memo;

    public bool PredictTheWinner(int[] nums)
    {
        int n = nums.Length;
        memo = new int?[n, n];

        return Dfs(nums, 0, n - 1) >= 0;
    }

    private int Dfs(int[] nums, int l, int r)
    {
        if (l == r)
            return nums[l];

        if (memo[l, r].HasValue)
            return memo[l, r].Value;

        int left = nums[l] - Dfs(nums, l + 1, r);
        int right = nums[r] - Dfs(nums, l, r - 1);

        memo[l, r] = Math.Max(left, right);
        return memo[l, r].Value;
    }
}

File: 2026/08-August/01-Saturday/solution.java
class Solution {
    public boolean PredictTheWinner(int[] nums) {
        int n = nums.length;
        Integer[][] memo = new Integer[n][n];
        return dfs(nums, 0, n - 1, memo) >= 0;
    }

    private int dfs(int[] nums, int left, int right, Integer[][] memo) {
        if (left == right) {
            return nums[left];
        }

        if (memo[left][right] != null) {
            return memo[left][right];
        }

        int takeLeft = nums[left] - dfs(nums, left + 1, right, memo);
        int takeRight = nums[right] - dfs(nums, left, right - 1, memo);

        return memo[left][right] = Math.max(takeLeft, takeRight);
    }
}
